src/ros2cs/ros2cs_core/executors/ManualExecutor.cs

```diff
@@ -0,0 +1,445 @@
+        {
+            this.IsIdle.Reset();
+            try
+            {
+                // check after resetting IsIdle to
+                // prevent race condition
+                if (this.RescanScheduled)
+                {
+                    return false;
+                }
+                if (this.WaitSet.TryWait(timeout, out var result))
+                {
+                    foreach (IWaitable waitable in result)
+                    {
+                        waitable.TryProcess();
+                    }
+                }
+            }
+            finally
+            {
+                this.IsIdle.Set();
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Rescan the nodes of this executor for
+        /// new objects to wait for and clear scheduled rescans.
+        /// </summary>
+        /// <remarks>
+        /// This method is thread safe if it itself or <see cref="TrySpin"/> is not executed concurrently
+        /// and enumerating the primitives of the nodes is thread safe.
+        /// </remarks>
+        public void Rescan()
+        {
+            // clear the wait set first to
+            // assert that the removed objects
+            // can be disposed even on error
+            this.WaitSet.Subscriptions.Clear();
+            this.WaitSet.Services.Clear();
+            this.WaitSet.Clients.Clear();
+            // clear the flag to prevent clearing rescans
+            // scheduled just after we finished rescaning
+            this.RescanScheduled = false;
+            try
+            {
+                // use the thread safe GetEnumerator wrapper
+                foreach (INode node in this)
+                {
+                    foreach (ISubscriptionBase subscription in node.Subscriptions)
+                    {
+                        this.WaitSet.Subscriptions.Add(subscription);
+                    }
+                    foreach (IServiceBase service in node.Services)
+                    {
+                        this.WaitSet.Services.Add(service);
+                    }
+                    foreach (IClientBase client in node.Clients)
+                    {
+                        this.WaitSet.Clients.Add(client);
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                // schedule rescan since the wait set may not be filled completely
+                this.ScheduleRescan();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Utility which handles automatic rescaning.
+        /// </summary>
+        /// <param name="timeout"> Maximum time to wait for work to become available. </param>
+        /// <returns>
+        /// <see cref="IEnumerator{bool}"/> trying to spin in each iteration
+        /// and yielding if a rescan had to be performed instead.
+        /// </returns>
+        public IEnumerator<bool> Spin(TimeSpan timeout)
```

Comment:
Not a huge fan of the IEnumerator approach here :thinking: 

It would be more user-friendly to use a spin more naturally, like `executor.Spin()` rather than wrapping the `IEnumerator` with `for`. A default timeout could also be nice.

Reply:
The problem with `Spin(TimeSpan)` is that it would be uninterruptible since disposing can only be done when the executor is not in use.
How about `SpinWhile(Func<bool> condition, TimeSpan timeout)`?

Reply:
Default arguments for `TimeSpan` are tricky which is why I created a overload which creates a default timeout.